crates/engine/src/operation/solve/solver/adapter.rs

```diff
@@ -142,14 +147,102 @@ impl<'a> query_solver::Operation for OperationAdapter<'a> {
 
     fn field_query_position(&self, field_id: Self::FieldId) -> usize {
         match &self.operation[field_id] {
-            BoundField::TypeName(field) => usize::from(field.key.query_position.unwrap()),
-            BoundField::Query(field) => usize::from(field.key.query_position.unwrap()),
+            BoundField::TypeName(field) => usize::from(field.query_position),
+            BoundField::Query(field) => usize::from(field.query_position),
             BoundField::Extra(_) => QueryPosition::EXTRA,
         }
     }
 }
 
 impl<'a> OperationAdapter<'a> {
+    /// Generate a new response key for a field with an distinct type, but compatible, than the super-graph.
+    /// This can happen when a subgraph A defines:
+    ///
+    /// ```ignore,graphql
+    /// type User {
+    ///   id: ID @shareable
+    /// }
+    /// ```
+    ///
+    /// and subgraph B defines:
+    ///
+    /// ```ignore,graphql
+    /// type User @key(fields: "id") {
+    ///   id: ID!
+    /// }
+    /// type Admin {
+    ///   id: ID
+    /// }
+    /// ```
+    ///
+    /// In this case the super-graph will use `id: ID`. A problem arises if we query something
+    /// like:
+    /// ```ignore,graphql
+    /// {
+    ///   ... on User {
+    ///     id
+    ///     name
+    ///   }
+    ///   ... on Admin {
+    ///     id
+    ///     name
+    ///   }
```

Comment:
I said it on slack, but this doesn't look like a legit query given the types that you've outlined.  If they're actually meant to be in a union/interface together can you make that clear.  And if not, can you eexplain how it works please.